cirq-core/cirq/circuits/circuit_operation.py

```diff
@@ -518,14 +543,19 @@ def with_measurement_key_mapping(self, key_map: Dict[str, str]) -> 'CircuitOpera
                 keys than this operation.
         """
         new_map = {}
-        for k_obj in self.circuit.all_measurement_key_objs():
+        for k_obj in protocols.measurement_keys_touched(self.circuit):
             k = k_obj.name
             k_new = self.measurement_key_map.get(k, k)
             k_new = key_map.get(k_new, k_new)
             if k_new != k:
                 new_map[k] = k_new
         new_op = self.replace(measurement_key_map=new_map)
-        if len(new_op._measurement_key_objs_()) != len(self._measurement_key_objs_()):
+        if (
+            len(protocols.measurement_key_objs(new_op)) != len(protocols.measurement_key_objs(self))
+            or len(protocols.control_keys(new_op)) != len(protocols.control_keys(self))
+            or len(protocols.measurement_keys_touched(new_op))
+            != len(protocols.measurement_keys_touched(self))
```

Comment:
I'm pretty sure we need it, to guarantee that control keys and measurement keys don't get remapped to each other.

In fact, maybe it's the *only* one we need, since if two measurement keys get mapped to a single measurement key, then that would also reduce the measurement_keys_touched. And there's no way that a single control key can get mapped to two control keys to compensate. So just checking measurement_keys_touched should be enough here. Can you sanity check that logic?

Reply:
Ah, my equation is wrong. `measurement_keys_touched` is a set-union, which would change size if two distinct keys of either type map to the same key.

As long as we always map measurement and control keys simultaneously, I believe you're correct - we can reduce to just the `measurement_keys_touched` check.